src/mission.rs

```diff
@@ -6,24 +6,37 @@ use crate::geo::orientation::Orientation;
 use crate::robot::Command;
 use crate::robot::Robot;
 
-#[derive(Debug, Eq, PartialEq)]
-pub struct Mission {
+#[derive(Debug, PartialEq)]
+pub enum Outcome {
+    Success(Robot),
+    Lost(Robot),
+}
+
+pub struct Mission<I, X>
+where
+    I: Iterator<Item = X>,
+{
     pub upper_right: Point,
+    source: I,
     scents: HashMap<Point, HashSet<Orientation>>,
 }
 
 const ORIGIN: Point = Point { x: 0, y: 0 };
 
-impl Mission {
-    pub fn new(upper_right: Point) -> Mission {
+impl<I, SourceItem> Mission<I, SourceItem>
+where
+    I: Iterator<Item = SourceItem>,
+{
+    pub fn new(upper_right: Point, source: I) -> Mission<I, SourceItem> {
```

Comment:
I did think about that, but then I'd need a trait object to hold the `source` in the `Mission` which always feels like admitting defeat :) So in the end, I went to look at how the iterators in the standard library are implemented and used generics instead. 

I would hope the errors would say something like "Mission does not implement Iterator trait with this Item type", but will admit not trying. 

This lead me down a deep rabbit hole thinking about how some monadic, FRP style stream combinators might work for iterators. Needless to say I didn't figure that out... but I'd love to have a `.and_then` or `.map_ok`, which would avoid the two specific iterator implementations.

Reply:
I got curious and checked if this gives good errors, and looks like it doesn't in the general case: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c7c07a5ab89bc7c920dc08f9a13f791b

Though also tried it on your code, and seems there's enough other annotations dotted about that your unliklely to hit that case.

Reply:
The message it gives
> the trait `std::iter::Iterator` is not implemented for `TestIterator<{integer}> 
seems reasonable enough to me, but I guess it could be confusing

Reply:
It's not an unreasonable error - just might have hoped for it to point towards the implementations of Iterator that do exist for TestIterator

Reply:
Ah yes, that would be nice :)

Reply:
Wanted to play about & see if my suggestion did need trait objects - ended up changing more than I'd planned, but think this is the kind of thing I was getting up: https://github.com/obmarg/martian-robots-rs/commit/f7ed1f6adb6bd9d1dd0a7441878e249d75fe4eb1

Reply:
That does looks pretty neat!